File: implementations/go-nethttp/filter.go
package ashurbanipal

import (
	"encoding/json"
	"fmt"
	"strings"
)

// MaxFilterBytes bounds the URL-decoded filter JSON (spec/protocol.md §5.4.2).
const MaxFilterBytes = 8192

// MaxConditions caps conditions per request (spec/protocol.md §5.4.2).
const MaxConditions = 10

// Wire operators must come from this allow-list before reaching SQL (spec/protocol.md §5.4.2).
var validOps = map[string]bool{
	"=": true, "!=": true, ">": true, "<": true, ">=": true, "<=": true,
	"LIKE": true, "ILIKE": true, "IS NULL": true, "IS NOT NULL": true,
}

func opTakesValue(op string) bool {
	return op != "IS NULL" && op != "IS NOT NULL"
}

// Condition holds untrusted filter input; BuildWhereClause validates Column
// against the live schema before SQL (spec/protocol.md §5.4.2).
type Condition struct {
	Logic  *string `json:"logic,omitempty"`
	Not    bool    `json:"not,omitempty"`
	Column string  `json:"column"`
	Op     string  `json:"op"`
	Value  *string `json:"value,omitempty"`
}

type FilterError struct {
	Reason string
}

func (e *FilterError) Error() string { return "invalid filter: " + e.Reason }

func filterErr(format string, args ...interface{}) *FilterError {
	return &FilterError{Reason: fmt.Sprintf(format, args...)}
}

// ParseFilter validates the JSON AST (spec/protocol.md §5.4.2); the frontend
// parses DSL text (spec/filter-dsl.md).
func ParseFilter(raw string) ([]Condition, error) {
	if len(raw) > MaxFilterBytes {
		return nil, filterErr("filter too long: %d bytes (max %d)", len(raw), MaxFilterBytes)
	}

	decoder := json.NewDecoder(strings.NewReader(raw))
	decoder.DisallowUnknownFields()
	var conditions []Condition
	if err := decoder.Decode(&conditions); err != nil {
		return nil, filterErr("filter must be a JSON array of conditions: %s", err)
	}
	if decoder.More() {
		return nil, filterErr("filter must be a single JSON array of conditions")
	}

	if len(conditions) > MaxConditions {
		return nil, filterErr("too many conditions: %d (max %d)", len(conditions), MaxConditions)
	}

	for i, cond := range conditions {
		if i == 0 && cond.Logic != nil {
			return nil, filterErr("logic must be absent on the first condition")
		}
		if i > 0 && cond.Logic == nil {
			return nil, filterErr("condition %d is missing logic (\"AND\" or \"OR\")", i)
		}
		if cond.Logic != nil && *cond.Logic != "AND" && *cond.Logic != "OR" {
			return nil, filterErr("condition %d has invalid logic %q", i, *cond.Logic)
		}
		if !validOps[cond.Op] {
			return nil, filterErr("condition %d has invalid op %q", i, cond.Op)
		}
		takesValue := opTakesValue(cond.Op)
		if takesValue && cond.Value == nil {
			return nil, filterErr("op %q requires a value", cond.Op)
		}
		if !takesValue && cond.Value != nil {
			return nil, filterErr("op %q takes no value", cond.Op)
		}
	}
	return conditions, nil
}

// opSQL returns a literal allow-listed SQL fragment, never wire text (spec/protocol.md §5.4.2).
func opSQL(op string) string {
	switch op {
	case "=":
		return "="
	case "!=":
		return "!="
	case ">":
		return ">"
	case "<":
		return "<"
	case ">=":
		return ">="
	case "<=":
		return "<="
	case "LIKE":
		return "LIKE"
	case "ILIKE":
		return "ILIKE"
	case "IS NULL":
		return "IS NULL"
	case "IS NOT NULL":
		return "IS NOT NULL"
	default:
		panic("opSQL called with an op BuildWhereClause's validOps check should have rejected: " + op)
	}
}

// BuildWhereClause renders conditions into a " where ..." SQL fragment
// with $N placeholders (numbered from $3: $1/$2 are reserved for
// limit/offset by QueryTable's caller) and the ordered bind values. Every
// column is matched against columnNames (the live information_schema
// allow-list) before being spliced in — the same discipline `sort` gets
// (spec/protocol.md §6). Conditions are joined by their own logic tokens,
// relying on SQL's native AND-tighter-than-OR precedence; there is no
// grouping/nesting in the AST.
func BuildWhereClause(conditions []Condition, columnNames []string) (string, []string, error) {
	if len(conditions) == 0 {
		return "", nil, nil
	}

	allowed := make(map[string]bool, len(columnNames))
	for _, c := range columnNames {
		allowed[c] = true
	}

	var values []string
	var clause strings.Builder
	nextParam := 3
	for i, cond := range conditions {
		if !allowed[cond.Column] {
			return "", nil, &NotAllowedError{What: fmt.Sprintf("column %q", cond.Column)}
		}
		// Defense in depth: BuildWhereClause is exported, so a future
		// caller could feed it conditions straight from JSON without
		// going through ParseFilter first — re-checking here is what
		// makes opSQL's hardcoded table load-bearing, not decorative
		// (PORTING.md hardening item 6).
		if !validOps[cond.Op] {
			return "", nil, filterErr("condition %d has invalid op %q", i, cond.Op)
		}

		quotedColumn := quoteIdent(cond.Column)
		var inner string
		if opTakesValue(cond.Op) {
			if cond.Value == nil {
				return "", nil, filterErr("op %q requires a value", cond.Op)
			}
			inner = fmt.Sprintf("%s::text %s $%d", quotedColumn, opSQL(cond.Op), nextParam)
			values = append(values, *cond.Value)
			nextParam++
		} else {
			inner = fmt.Sprintf("%s::text %s", quotedColumn, opSQL(cond.Op))
		}

		var wrapped string
		if cond.Not {
			wrapped = "(NOT (" + inner + "))"
		} else {
			wrapped = "(" + inner + ")"
		}

		if i > 0 {
			if cond.Logic == nil {
				return "", nil, filterErr("condition %d is missing logic", i)
			}
			if *cond.Logic == "OR" {
				clause.WriteString(" OR ")
			} else {
				clause.WriteString(" AND ")
			}
		}
		clause.WriteString(wrapped)
	}
	return " where " + clause.String(), values, nil
}

File: implementations/node-express/src/filter.ts
import { FilterError, NotAllowedError, quoteIdent } from "./errors.js";

/** Maximum URL-decoded filter JSON bytes (spec/protocol.md §5.4.2). */
export const MAX_FILTER_BYTES = 8192;

/** Maximum conditions per filter request (spec/protocol.md §5.4.2). */
export const MAX_CONDITIONS = 10;

// Wire operators must come from this allow-list before reaching SQL (spec/protocol.md §5.4.2).
const VALID_OPS = new Set(["=", "!=", ">", "<", ">=", "<=", "LIKE", "ILIKE", "IS NULL", "IS NOT NULL"]);

function opTakesValue(op: string): boolean {
  return op !== "IS NULL" && op !== "IS NOT NULL";
}

/** Untrusted filter input; buildWhereClause validates `column` before SQL (spec/protocol.md §5.4.2). */
export interface Condition {
  logic?: "AND" | "OR";
  not?: boolean;
  column: string;
  op: string;
  value?: string;
}

/** Validates JSON ASTs (spec/protocol.md §5.4.2); the frontend parses DSL text (spec/filter-dsl.md). */
export function parseFilter(raw: string): Condition[] {
  const byteLength = Buffer.byteLength(raw, "utf8");
  if (byteLength > MAX_FILTER_BYTES) {
    throw new FilterError(`filter too long: ${byteLength} bytes (max ${MAX_FILTER_BYTES})`);
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch (err) {
    throw new FilterError(
      `filter must be a JSON array of conditions: ${err instanceof Error ? err.message : String(err)}`,
    );
  }
  if (!Array.isArray(parsed)) {
    throw new FilterError("filter must be a JSON array of conditions");
  }
  if (parsed.length > MAX_CONDITIONS) {
    throw new FilterError(`too many conditions: ${parsed.length} (max ${MAX_CONDITIONS})`);
  }

  const conditions: Condition[] = parsed.map((raw, i) => validateConditionShape(raw, i));

  conditions.forEach((cond, i) => {
    if (i === 0 && cond.logic !== undefined) {
      throw new FilterError("logic must be absent on the first condition");
    }
    if (i > 0 && cond.logic === undefined) {
      throw new FilterError(`condition ${i} is missing logic ("AND" or "OR")`);
    }
    if (cond.logic !== undefined && cond.logic !== "AND" && cond.logic !== "OR") {
      throw new FilterError(`condition ${i} has invalid logic "${String(cond.logic)}"`);
    }
    if (!VALID_OPS.has(cond.op)) {
      throw new FilterError(`condition ${i} has invalid op "${cond.op}"`);
    }
    const takesValue = opTakesValue(cond.op);
    if (takesValue && cond.value === undefined) {
      throw new FilterError(`op "${cond.op}" requires a value`);
    }
    if (!takesValue && cond.value !== undefined) {
      throw new FilterError(`op "${cond.op}" takes no value`);
    }
  });

  return conditions;
}

// Positional validation stays in parseFilter because it sees the full array.
function validateConditionShape(raw: unknown, index: number): Condition {
  if (typeof raw !== "object" || raw === null || Array.isArray(raw)) {
    throw new FilterError(`condition ${index} must be a JSON object`);
  }
  const obj = raw as Record<string, unknown>;
  const allowedKeys = new Set(["logic", "not", "column", "op", "value"]);
  for (const key of Object.keys(obj)) {
    if (!allowedKeys.has(key)) {
      throw new FilterError(`condition ${index} has unknown field "${key}"`);
    }
  }
  if (typeof obj.column !== "string") {
    throw new FilterError(`condition ${index} is missing a string "column"`);
  }
  if (typeof obj.op !== "string") {
    throw new FilterError(`condition ${index} is missing a string "op"`);
  }
  if (obj.logic !== undefined && typeof obj.logic !== "string") {
    throw new FilterError(`condition ${index} has a non-string "logic"`);
  }
  if (obj.not !== undefined && typeof obj.not !== "boolean") {
    throw new FilterError(`condition ${index} has a non-boolean "not"`);
  }
  if (obj.value !== undefined && typeof obj.value !== "string") {
    throw new FilterError(`condition ${index} has a non-string "value"`);
  }
  return {
    logic: obj.logic as "AND" | "OR" | undefined,
    not: (obj.not as boolean | undefined) ?? false,
    column: obj.column,
    op: obj.op,
    value: obj.value as string | undefined,
  };
}

// opSql returns a literal allow-listed SQL fragment, never wire text (spec/protocol.md §5.4.2).
function opSql(op: string): string {
  switch (op) {
    case "=":
      return "=";
    case "!=":
      return "!=";
    case ">":
      return ">";
    case "<":
      return "<";
    case ">=":
      return ">=";
    case "<=":
      return "<=";
    case "LIKE":
      return "LIKE";
    case "ILIKE":
      return "ILIKE";
    case "IS NULL":
      return "IS NULL";
    case "IS NOT NULL":
      return "IS NOT NULL";
    default:
      throw new Error(`opSql called with an op VALID_OPS should have rejected: ${op}`);
  }
}

export interface WhereClause {
  where: string;
  values: string[];
}

/**
 * Renders conditions into a " where ..." SQL fragment with $N placeholders
 * (numbered starting at startParam, since callers reserve earlier $N slots
 * for limit/offset) and the ordered bind values. Every column is matched
 * against columnNames (the live information_schema allow-list) before
 * being spliced in — the same discipline `sort` gets (spec/protocol.md
 * §6). Conditions are joined by their own logic tokens, relying on SQL's
 * native AND-tighter-than-OR precedence; there is no grouping/nesting in
 * the AST.
 */
export function buildWhereClause(conditions: Condition[], columnNames: string[], startParam: number): WhereClause {
  if (conditions.length === 0) {
    return { where: "", values: [] };
  }

  const allowed = new Set(columnNames);
  const values: string[] = [];
  let clause = "";
  let nextParam = startParam;

  conditions.forEach((cond, i) => {
    if (!allowed.has(cond.column)) {
      throw new NotAllowedError(`column "${cond.column}"`);
    }
    // Defense in depth: buildWhereClause is only ever fed parseFilter's
    // already-op-validated output in production, but it's an exported
    // function a future caller (or a test) could feed conditions to
    // directly — re-checking here is what makes opSql's hardcoded table
    // load-bearing rather than decorative.
    if (!VALID_OPS.has(cond.op)) {
      throw new FilterError(`condition ${i} has invalid op "${cond.op}"`);
    }

    const quotedColumn = quoteIdent(cond.column);
    let inner: string;
    if (opTakesValue(cond.op)) {
      if (cond.value === undefined) {
        throw new FilterError(`op "${cond.op}" requires a value`);
      }
      inner = `${quotedColumn}::text ${opSql(cond.op)} $${nextParam}`;
      values.push(cond.value);
      nextParam++;
    } else {
      inner = `${quotedColumn}::text ${opSql(cond.op)}`;
    }

    const wrapped = cond.not ? `(NOT (${inner}))` : `(${inner})`;

    if (i > 0) {
      if (cond.logic === undefined) {
        throw new FilterError(`condition ${i} is missing logic`);
      }
      clause += cond.logic === "OR" ? " OR " : " AND ";
    }
    clause += wrapped;
  });

  return { where: ` where ${clause}`, values };
}
